react_basic/src/pages/PageWrapper.jsx

```diff
@@ -0,0 +1,154 @@
+import { createContext, useEffect, useState } from 'react';
+import { Outlet, useLocation, useNavigate } from 'react-router-dom';
+import Header from '../components/Header.jsx';
+import Footer from '../components/Footer.jsx';
+import PopupMessage from '../components/PopupMessage.jsx';
+import FullscreenPopup from '../components/FullscreenPopup.jsx';
+import Advert from '../components/Advert.jsx';
+import themeService from '../services/themeService.js';
+import localStorageService from '../services/localStorageService.js';
+import urlUtils from '../utils/urlUtils.js';
+import ErrorBoundary from '../components/ErrorBoundary.jsx';
+import ErrorMsg from '../components/ErrorMsg.jsx';
+import '../styles/PageWrapper.css';
+
+const PageWrapperContext = createContext();
+
+function PageWrapper() {
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  const [user, setUser] = useState(localStorageService.getLoggedInUser());
+  const [theme, setTheme] = useState();
+  const [favorites, setFavorites] = useState([]);
+
+  const [successMsg, setSuccessMsg] = useState();
+  const [successMsgMillis, setSuccessMsgMillis] = useState();
+  const [successMsgCallback, setSuccessMsgCallback] = useState();
+
+  const [errorMsg, setErrorMsg] = useState();
+  const [errorMsgMillis, setErrorMsgMillis] = useState();
+  const [errorMsgCallback, setErrorMsgCallback] = useState();
+  const [fullscreenPopupContent, setFullscreenPopupContent] = useState();
+
+  useEffect(() => {
+    document.body.addEventListener('keydown', (e) => {
+      if (e.key === 'B' && e.shiftKey && e.ctrlKey && e.altKey) {
+        e.stopImmediatePropagation();
+
+        navigate('/billy');
+      }
+    });
+  }, []);
+
+  function successPopupResetCallback() {
```

Comment:
From my side of view such complex and heavy functions are very difficult to understand and read. Also this function contain double stIsLoading(true) calls, but this is inefficient. It's better to divide it into logical functions:
![image](https://github.com/user-attachments/assets/7e6aecd1-3f01-4e55-8ad2-1763bea529d1)